File: ex/test/gralkor/client/http_test.exs
defmodule Gralkor.Client.HTTPTest do
  use ExUnit.Case, async: false

  alias Gralkor.Client.HTTP

  setup do
    Req.Test.set_req_test_to_private(self())
    :ok
  end

  # ── Backend configuration for the shared port contract ───────────────

  defp configure_backend(:recall, {:ok, block}) do
    stub(fn conn ->
      body = expect_post(conn, "/recall")
      assert %{"group_id" => _, "query" => _} = body
      wire_body = if is_nil(block), do: "", else: block
      Req.Test.json(conn, %{"memory_block" => wire_body})
    end)
  end

  defp configure_backend(:recall, {:error, _reason}) do
    stub(fn conn ->
      _ = expect_post(conn, "/recall")
      Plug.Conn.send_resp(conn, 503, "")
    end)
  end

  defp configure_backend(:capture, :ok) do
    stub(fn conn ->
      body = expect_post(conn, "/capture")

      assert %{
               "session_id" => _,
               "group_id" => _,
               "messages" => messages
             } = body

      assert is_list(messages)

      for msg <- messages do
        assert %{"role" => role, "content" => content} = msg
        assert role in ["user", "assistant", "behaviour"]
        assert is_binary(content)
      end

      Plug.Conn.send_resp(conn, 204, "")
    end)
  end

  defp configure_backend(:capture, {:error, _}) do
    stub(fn conn ->
      _ = expect_post(conn, "/capture")
      Plug.Conn.send_resp(conn, 500, "")
    end)
  end

  defp configure_backend(:memory_search, {:ok, text}) do
    stub(fn conn ->
      body = expect_post(conn, "/tools/memory_search")
      assert %{"group_id" => _, "session_id" => _, "query" => _} = body
      Req.Test.json(conn, %{"text" => text})
    end)
  end

  defp configure_backend(:memory_search, {:error, _}) do
    stub(fn conn ->
      _ = expect_post(conn, "/tools/memory_search")
      Plug.Conn.send_resp(conn, 502, "")
    end)
  end

  defp configure_backend(:end_session, :ok) do
    stub(fn conn ->
      body = expect_post(conn, "/session_end")
      assert %{"session_id" => _} = body
      Plug.Conn.send_resp(conn, 204, "")
    end)
  end

  defp configure_backend(:end_session, {:error, _}) do
    stub(fn conn ->
      _ = expect_post(conn, "/session_end")
      Plug.Conn.send_resp(conn, 500, "")
    end)
  end

  defp configure_backend(:memory_add, :ok) do
    stub(fn conn ->
      body = expect_post(conn, "/tools/memory_add")
      assert %{"group_id" => _, "content" => _} = body
      Req.Test.json(conn, %{"status" => "stored"})
    end)
  end

  defp configure_backend(:memory_add, {:error, _}) do
    stub(fn conn ->
      _ = expect_post(conn, "/tools/memory_add")
      Plug.Conn.send_resp(conn, 500, "")
    end)
  end

  defp configure_backend(:health_check, :ok) do
    stub(fn conn ->
      assert conn.method == "GET"
      assert conn.request_path == "/health"
      Req.Test.json(conn, %{"status" => "ok"})
    end)
  end

  defp configure_backend(:health_check, {:error, _}) do
    stub(fn conn -> Plug.Conn.send_resp(conn, 503, "") end)
  end

  defp configure_backend(:build_indices, {:ok, %{status: status}}) do
    stub(fn conn ->
      assert conn.method == "POST"
      assert conn.request_path == "/build-indices"
      Req.Test.json(conn, %{"status" => status})
    end)
  end

  defp configure_backend(:build_indices, {:error, _}) do
    stub(fn conn ->
      assert conn.request_path == "/build-indices"
      Plug.Conn.send_resp(conn, 503, "")
    end)
  end

  defp configure_backend(:build_communities, {:ok, %{communities: c, edges: e}}) do
    stub(fn conn ->
      body = expect_post(conn, "/build-communities")
      assert %{"group_id" => _} = body
      Req.Test.json(conn, %{"communities" => c, "edges" => e})
    end)
  end

  defp configure_backend(:build_communities, {:error, _}) do
    stub(fn conn ->
      _ = expect_post(conn, "/build-communities")
      Plug.Conn.send_resp(conn, 503, "")
    end)
  end

  use Gralkor.ClientContract, client: HTTP

  # ── Adapter-specific behaviour ───────────────────────────────────────

  describe "every HTTP request" do
    test "carries no Authorization header" do
      parent = self()

      stub(fn conn ->
        send(parent, {:auth_header, Plug.Conn.get_req_header(conn, "authorization")})
        Req.Test.json(conn, %{"memory_block" => ""})
      end)

      {:ok, nil} = HTTP.recall("g1", "s1", "q")

      assert_receive {:auth_header, []}
    end
  end

  describe "if Gralkor responds with a non-2xx status" do
    test "{:error, {:http_status, status, body}} is returned" do
      stub(fn conn -> Plug.Conn.send_resp(conn, 418, "i'm a teapot") end)

      assert {:error, {:http_status, 418, "i'm a teapot"}} = HTTP.recall("g1", "s1", "q")
    end
  end

  describe "if the app env is missing" do
    test "the call raises" do
      previous = Application.get_env(:gralkor_ex, :client_http)
      Application.delete_env(:gralkor_ex, :client_http)

      try do
        assert_raise ArgumentError, fn -> HTTP.recall("g1", "s1", "q") end
      after
        Application.put_env(:gralkor_ex, :client_http, previous)
      end
    end
  end

  describe "when recall is called with a non-blank string session_id" do
    test "the session_id field is included in the HTTP body" do
      parent = self()

      stub(fn conn ->
        body = expect_post(conn, "/recall")
        send(parent, {:body, body})
        Req.Test.json(conn, %{"memory_block" => ""})
      end)

      assert {:ok, nil} = HTTP.recall("g1", "s1", "q")

      assert_receive {:body, body}
      assert body["session_id"] == "s1"
    end
  end

  describe "when recall is called with a nil session_id" do
    test "the session_id field is omitted from the HTTP body" do
      parent = self()

      stub(fn conn ->
        body = expect_post(conn, "/recall")
        send(parent, {:body, body})
        Req.Test.json(conn, %{"memory_block" => ""})
      end)

      assert {:ok, nil} = HTTP.recall("g1", nil, "q")

      assert_receive {:body, body}
      refute Map.has_key?(body, "session_id")
    end
  end

  describe "capture/3 wire shape" do
    test "serialises Gralkor.Message structs to {role, content} JSON" do
      parent = self()

      stub(fn conn ->
        body = expect_post(conn, "/capture")
        send(parent, {:body, body})
        Plug.Conn.send_resp(conn, 204, "")
      end)

      messages = [
        Gralkor.Message.new("user", "hello"),
        Gralkor.Message.new("behaviour", "thought: considering"),
        Gralkor.Message.new("assistant", "hi!")
      ]

      assert :ok = HTTP.capture("s1", "g1", messages)

      assert_receive {:body, body}

      assert body["messages"] == [
               %{"role" => "user", "content" => "hello"},
               %{"role" => "behaviour", "content" => "thought: considering"},
               %{"role" => "assistant", "content" => "hi!"}
             ]
    end
  end

  describe "when the transport fails with a connection-level error, then the call is retried exactly once, when the retry succeeds" do
    test "the response is returned normally" do
      parent = self()

      Req.Test.expect(:gralkor_stub, fn conn ->
        send(parent, :stub_called)
        Req.Test.transport_error(conn, :closed)
      end)

      Req.Test.expect(:gralkor_stub, fn conn ->
        send(parent, :stub_called)
        Req.Test.json(conn, %{"memory_block" => ""})
      end)

      assert {:ok, nil} = HTTP.recall("g1", "s1", "q")

      assert_received :stub_called
      assert_received :stub_called
      refute_received :stub_called
    end
  end

  describe "when the transport fails with a connection-level error, then the call is retried exactly once, when the retry also fails" do
    test "the failure surfaces to the caller" do
      parent = self()

      Req.Test.expect(:gralkor_stub, 2, fn conn ->
        send(parent, :stub_called)
        Req.Test.transport_error(conn, :timeout)
      end)

      assert {:error, %Req.TransportError{reason: :timeout}} = HTTP.recall("g1", "s1", "q")

      assert_received :stub_called
      assert_received :stub_called
      refute_received :stub_called
    end
  end

  describe "when the server returns a non-2xx HTTP response (including 429)" do
    test "no retry is attempted — the response surfaces immediately" do
      parent = self()

      Req.Test.expect(:gralkor_stub, fn conn ->
        send(parent, :stub_called)
        Plug.Conn.send_resp(conn, 503, "")
      end)

      assert {:error, {:http_status, 503, _}} = HTTP.recall("g1", "s1", "q")

      assert_received :stub_called
      refute_received :stub_called
    end

    # The google-genai SDK owns Vertex-upstream retries (see
    # gralkor/TEST_TREES.md > Retry ownership). A 429 that reaches this
    # layer means the SDK has already exhausted its retries; retrying
    # here would only amplify load. 429 surfaces like any other non-2xx.
    test "429 surfaces without retry — the SDK at L6.5 owns Vertex-upstream retries" do
      parent = self()

      Req.Test.expect(:gralkor_stub, fn conn ->
        send(parent, :stub_called)

        conn
        |> Plug.Conn.put_resp_header("retry-after", "0")
        |> Plug.Conn.put_resp_content_type("application/json")
        |> Plug.Conn.send_resp(429, Jason.encode!(%{"detail" => "rate limited"}))
      end)

      assert {:error, {:http_status, 429, %{"detail" => "rate limited"}}} =
               HTTP.recall("g1", "s1", "q")

      assert_received :stub_called
      refute_received :stub_called
    end
  end

  describe "if the transport fails with any other error" do
    test "no retry is attempted — the failure surfaces immediately (fail-fast default)" do
      parent = self()

      Req.Test.expect(:gralkor_stub, fn conn ->
        send(parent, :stub_called)
        Req.Test.transport_error(conn, :econnrefused)
      end)

      assert {:error, %Req.TransportError{reason: :econnrefused}} = HTTP.recall("g1", "s1", "q")

      assert_received :stub_called
      refute_received :stub_called
    end
  end

  describe "if capture is called with a blank string session_id" do
    test "the call raises with ArgumentError" do
      messages = [Gralkor.Message.new("user", "q"), Gralkor.Message.new("assistant", "a")]

      assert_raise ArgumentError, ~r/session_id must be a non-blank string/, fn ->
        HTTP.capture("", "g1", messages)
      end
    end
  end

  describe "if capture is called with a nil session_id" do
    test "the call raises with ArgumentError" do
      messages = [Gralkor.Message.new("user", "q"), Gralkor.Message.new("assistant", "a")]

      assert_raise ArgumentError, ~r/session_id must be a non-blank string/, fn ->
        HTTP.capture(nil, "g1", messages)
      end
    end
  end

  describe "if memory_search is called with a blank string session_id" do
    test "the call raises with ArgumentError" do
      assert_raise ArgumentError, ~r/session_id must be a non-blank string/, fn ->
        HTTP.memory_search("g1", "", "q")
      end
    end
  end

  describe "if memory_search is called with a nil session_id" do
    test "the call raises with ArgumentError" do
      assert_raise ArgumentError, ~r/session_id must be a non-blank string/, fn ->
        HTTP.memory_search("g1", nil, "q")
      end
    end
  end

  describe "if end_session is called with a blank string session_id" do
    test "the call raises with ArgumentError" do
      assert_raise ArgumentError, ~r/session_id must be a non-blank string/, fn ->
        HTTP.end_session("")
      end
    end
  end

  describe "if end_session is called with a nil session_id" do
    test "the call raises with ArgumentError" do
      assert_raise ArgumentError, ~r/session_id must be a non-blank string/, fn ->
        HTTP.end_session(nil)
      end
    end
  end

  # ── Helpers ──────────────────────────────────────────────────────────

  defp stub(fun), do: Req.Test.stub(:gralkor_stub, fun)

  defp expect_post(conn, path) do
    assert conn.method == "POST"
    assert conn.request_path == path
    {:ok, raw, _conn} = Plug.Conn.read_body(conn)
    Jason.decode!(raw)
  end
end

File: ts/test/client/http.test.ts
import { describe, it, expect, beforeEach, afterEach, vi } from "vitest";
import { GralkorHttpClient } from "../../src/client/http.js";
import { gralkorClientContract } from "../contract/gralkor-client.contract.js";
import type { Result } from "../../src/client.js";

type StubKey =
  | "recall"
  | "capture"
  | "endSession"
  | "memorySearch"
  | "memoryAdd"
  | "healthCheck"
  | "buildIndices"
  | "buildCommunities";

type Stub = {
  status: number;
  body: unknown;
};

function makeStubbedClient(): {
  client: GralkorHttpClient;
  stub: (key: StubKey, response: Result<unknown>) => void;
  lastRequest: () => { url: string; init: RequestInit } | null;
} {
  const stubs = new Map<StubKey, Stub>();
  let last: { url: string; init: RequestInit } | null = null;

  const fetchStub: typeof fetch = async (input, init) => {
    const url = typeof input === "string" ? input : input.toString();
    last = { url, init: init ?? {} };

    // route based on path
    const path = new URL(url).pathname;
    const key: StubKey | null =
      path === "/recall" ? "recall" :
      path === "/capture" ? "capture" :
      path === "/session_end" ? "endSession" :
      path === "/tools/memory_search" ? "memorySearch" :
      path === "/tools/memory_add" ? "memoryAdd" :
      path === "/health" ? "healthCheck" :
      path === "/build-indices" ? "buildIndices" :
      path === "/build-communities" ? "buildCommunities" :
      null;

    if (!key) throw new Error(`no stub for path ${path}`);

    const s = stubs.get(key);
    if (!s) throw new Error(`no stub configured for ${key}`);

    return new Response(s.body === undefined ? null : JSON.stringify(s.body), {
      status: s.status,
      headers: { "content-type": "application/json" },
    });
  };

  const client = new GralkorHttpClient({ baseUrl: "http://gralkor.test", fetch: fetchStub });

  const stub = (key: StubKey, response: Result<unknown>) => {
    if ("ok" in response) {
      // Translate Result<T> → HTTP 2xx with body the HTTP adapter will decode back to { ok: T }
      const body: Record<string, unknown> =
        key === "recall" ? { memory_block: response.ok === null ? "" : response.ok } :
        key === "memorySearch" ? { text: response.ok } :
        key === "memoryAdd" ? { status: "stored" } :
        key === "buildIndices" ? (response.ok as Record<string, unknown>) :
        key === "buildCommunities" ? (response.ok as Record<string, unknown>) :
        {};
      stubs.set(key, {
        status: key === "capture" || key === "endSession" ? 204 : 200,
        body: key === "capture" || key === "endSession" ? undefined : body,
      });
    } else {
      stubs.set(key, { status: 503, body: response.error });
    }
  };

  return { client, stub, lastRequest: () => last };
}

describe("GralkorHttpClient (via shared contract)", () => {
  const ctx = { current: makeStubbedClient() };
  beforeEach(() => {
    ctx.current = makeStubbedClient();
  });

  gralkorClientContract({
    make: () => ctx.current.client,
    configureBackend: (_c, op, response) => ctx.current.stub(op, response),
  });
});

describe("GralkorHttpClient (adapter-specific)", () => {
  let harness: ReturnType<typeof makeStubbedClient>;

  beforeEach(() => {
    harness = makeStubbedClient();
  });

  describe("every HTTP request", () => {
    it("carries no Authorization header", async () => {
      harness.stub("recall", { ok: null });
      await harness.client.recall("g1", "s1", "q");
      const headers = new Headers(harness.lastRequest()?.init.headers);
      expect(headers.has("authorization")).toBe(false);
    });
  });

  describe("if Gralkor responds with a non-2xx status", () => {
    it("returns { error: { kind: 'http_status', ... } }", async () => {
      // Direct override — the stub() helper maps { error } to 503. We want an arbitrary non-2xx.
      const client = new GralkorHttpClient({
        baseUrl: "http://gralkor.test",
        fetch: async () => new Response("i'm a teapot", { status: 418 }),
      });
      const r = await client.recall("g1", "s1", "q");
      expect("error" in r).toBe(true);
      if ("error" in r) {
        expect((r.error as { kind: string }).kind).toBe("http_status");
      }
    });
  });

  describe("if session_id is blank", () => {
    it("recall/3 throws", async () => {
      await expect(harness.client.recall("g1", "", "q")).rejects.toThrow(/session_id/);
    });
    it("capture/3 throws", async () => {
      const messages = [{ role: "user" as const, content: "q" }];
      await expect(harness.client.capture("", "g1", messages)).rejects.toThrow(/session_id/);
    });
    it("memorySearch/3 throws", async () => {
      await expect(harness.client.memorySearch("g1", "", "q")).rejects.toThrow(/session_id/);
    });
    it("endSession/1 throws", async () => {
      await expect(harness.client.endSession("")).rejects.toThrow(/session_id/);
    });
  });

  describe("when the transport fails with a connection-level error, then the call is retried exactly once, when the retry succeeds", () => {
    it("the response is returned normally", async () => {
      let calls = 0;
      const client = new GralkorHttpClient({
        baseUrl: "http://gralkor.test",
        fetch: async () => {
          calls += 1;
          if (calls === 1) {
            const err = new Error("connection reset") as Error & { code?: string };
            err.code = "ECONNRESET";
            throw err;
          }
          return new Response(JSON.stringify({ memory_block: "" }), {
            status: 200,
            headers: { "content-type": "application/json" },
          });
        },
      });
      const result = await client.recall("g1", "s1", "q");
      expect(result).toEqual({ ok: null });
      expect(calls).toBe(2);
    });
  });

  describe("when the transport fails with a connection-level error, then the call is retried exactly once, when the retry also fails", () => {
    it("the failure surfaces to the caller", async () => {
      let calls = 0;
      const client = new GralkorHttpClient({
        baseUrl: "http://gralkor.test",
        fetch: async () => {
          calls += 1;
          const err = new Error("socket timeout") as Error & { code?: string };
          err.code = "ETIMEDOUT";
          throw err;
        },
      });
      const result = await client.recall("g1", "s1", "q");
      expect("error" in result).toBe(true);
      expect(calls).toBe(2);
    });
  });

  describe("when the server returns any HTTP response (including non-2xx)", () => {
    it("no retry is attempted — the response surfaces immediately", async () => {
      let calls = 0;
      const client = new GralkorHttpClient({
        baseUrl: "http://gralkor.test",
        fetch: async () => {
          calls += 1;
          return new Response("", { status: 503 });
        },
      });
      const result = await client.recall("g1", "s1", "q");
      expect("error" in result).toBe(true);
      if ("error" in result) {
        expect((result.error as { kind: string }).kind).toBe("http_status");
      }
      expect(calls).toBe(1);
    });
  });

  describe("if the transport fails with any other error", () => {
    it("no retry is attempted — the failure surfaces immediately (fail-fast default)", async () => {
      let calls = 0;
      const client = new GralkorHttpClient({
        baseUrl: "http://gralkor.test",
        fetch: async () => {
          calls += 1;
          const err = new Error("dns lookup failed") as Error & { code?: string };
          err.code = "ENOTFOUND";
          throw err;
        },
      });
      const result = await client.recall("g1", "s1", "q");
      expect("error" in result).toBe(true);
      if ("error" in result) {
        expect((result.error as { kind: string }).kind).toBe("network");
      }
      expect(calls).toBe(1);
    });
  });

  describe("when recall is called without maxResults", () => {
    it("omits max_results from the request body so the server applies its default", async () => {
      harness.stub("recall", { ok: null });
      await harness.client.recall("g1", "s1", "q");
      const body = JSON.parse(String(harness.lastRequest()?.init.body));
      expect(body).toEqual({ group_id: "g1", session_id: "s1", query: "q" });
    });
  });

  describe("when recall is called with maxResults", () => {
    it("includes max_results in the request body", async () => {
      harness.stub("recall", { ok: null });
      await harness.client.recall("g1", "s1", "q", 5);
      const body = JSON.parse(String(harness.lastRequest()?.init.body));
      expect(body).toEqual({ group_id: "g1", session_id: "s1", query: "q", max_results: 5 });
    });
  });

  describe("when memorySearch is called without max args", () => {
    it("omits both max_results and max_entity_results from the request body", async () => {
      harness.stub("memorySearch", { ok: "text" });
      await harness.client.memorySearch("g1", "s1", "q");
      const body = JSON.parse(String(harness.lastRequest()?.init.body));
      expect(body).toEqual({ group_id: "g1", session_id: "s1", query: "q" });
    });
  });

  describe("when memorySearch is called with maxResults and maxEntityResults", () => {
    it("includes both in the request body", async () => {
      harness.stub("memorySearch", { ok: "text" });
      await harness.client.memorySearch("g1", "s1", "q", 7, 3);
      const body = JSON.parse(String(harness.lastRequest()?.init.body));
      expect(body).toEqual({
        group_id: "g1",
        session_id: "s1",
        query: "q",
        max_results: 7,
        max_entity_results: 3,
      });
    });
  });
});

describe("client-timeouts", () => {
  beforeEach(() => {
    vi.useFakeTimers();
  });
  afterEach(() => {
    vi.useRealTimers();
  });

  const hangingFetch: typeof fetch = (_input, init) =>
    new Promise((_resolve, reject) => {
      const signal = init?.signal;
      if (signal?.aborted) {
        reject(new DOMException("aborted", "AbortError"));
        return;
      }
      signal?.addEventListener("abort", () =>
        reject(new DOMException("aborted", "AbortError")),
      );
    });

  const timed: Array<{
    path: string;
    ms: number;
    call: (c: GralkorHttpClient) => Promise<unknown>;
  }> = [
    { path: "/health", ms: 2_000, call: (c) => c.healthCheck() },
    { path: "/recall", ms: 5_000, call: (c) => c.recall("g", "s", "q") },
    {
      path: "/capture",
      ms: 5_000,
      call: (c) => c.capture("s", "g", [{ role: "user", content: "q" }]),
    },
    { path: "/session_end", ms: 5_000, call: (c) => c.endSession("s") },
    {
      path: "/tools/memory_search",
      ms: 10_000,
      call: (c) => c.memorySearch("g", "s", "q"),
    },
    {
      path: "/tools/memory_add",
      ms: 60_000,
      call: (c) => c.memoryAdd("g", "content", null),
    },
  ];

  for (const { path, ms, call } of timed) {
    it(`${path} aborts at ${ms}ms`, async () => {
      const client = new GralkorHttpClient({
        baseUrl: "http://gralkor.test",
        fetch: hangingFetch,
      });
      const promise = call(client);

      await vi.advanceTimersByTimeAsync(ms + 10);
      const r = (await promise) as { error?: { kind: string } };
      expect(r.error?.kind).toBe("network");
    });
  }

  const admin: Array<{
    path: string;
    call: (c: GralkorHttpClient) => Promise<unknown>;
    body: unknown;
    expected: unknown;
  }> = [
    {
      path: "/build-indices",
      call: (c) => c.buildIndices(),
      body: { status: "ok" },
      expected: { ok: { status: "ok" } },
    },
    {
      path: "/build-communities",
      call: (c) => c.buildCommunities("g"),
      body: { communities: 3, edges: 17 },
      expected: { ok: { communities: 3, edges: 17 } },
    },
  ];

  for (const { path, call, body, expected } of admin) {
    it(`${path} has no client-side deadline`, async () => {
      let resolveFetch!: (r: Response) => void;
      const controlledFetch: typeof fetch = (_input, init) =>
        new Promise((resolve, reject) => {
          const signal = init?.signal;
          if (signal?.aborted) {
            reject(new DOMException("aborted", "AbortError"));
            return;
          }
          signal?.addEventListener("abort", () =>
            reject(new DOMException("aborted", "AbortError")),
          );
          resolveFetch = resolve;
        });

      const client = new GralkorHttpClient({
        baseUrl: "http://gralkor.test",
        fetch: controlledFetch,
      });
      const promise = call(client);

      await vi.advanceTimersByTimeAsync(60 * 60 * 1000);
      await Promise.resolve();

      resolveFetch(
        new Response(JSON.stringify(body), {
          status: 200,
          headers: { "content-type": "application/json" },
        }),
      );
      const r = await promise;
      expect(r).toEqual(expected);
    });
  }
});
